File: android/knowledge-workbench/app/src/main/java/com/mllm/knowledgeworkbench/data/KnowledgeContracts.java
package com.mllm.knowledgeworkbench.data;

public final class KnowledgeContracts {
    private KnowledgeContracts() {}

    public enum SearchMode {
        LEXICAL,
        EMBEDDING,
        HYBRID
    }

    public record SearchHit(
        String documentId,
        String chunkId,
        String sourceUri,
        String title,
        int ordinal,
        String excerpt,
        double score
    ) {
        public SearchHit withScore(double newScore) {
            return new SearchHit(documentId, chunkId, sourceUri, title, ordinal, excerpt, newScore);
        }
    }

    public record Snapshot(
        boolean lexicalReady,
        boolean fts5Ready,
        String lexicalMode,
        String providerId,
        String modelId,
        int totalChunks,
        int indexedChunks,
        String databasePath
    ) {
        public boolean hybridReady() {
            return lexicalReady && totalChunks > 0 && totalChunks == indexedChunks;
        }
    }

    public record IndexProgress(
        int completed,
        int total,
        String currentChunkId
    ) {
        public int percent() {
            if (total <= 0) return 0;
            return (int) Math.round((completed * 100.0d) / total);
        }
    }

    @FunctionalInterface
    public interface ProgressListener {
        void onProgress(IndexProgress progress);
    }
}

File: src/MLLM.Workbench.Knowledge/KnowledgeContracts.cs
namespace MLLM.Workbench.Knowledge;

public sealed record KnowledgeStoreOptions(string DatabasePath);

public sealed record KnowledgeChunk(
    string ChunkId,
    int Ordinal,
    string Content);

public sealed record KnowledgeDocument(
    string DocumentId,
    string SourceUri,
    string Title,
    IReadOnlyList<KnowledgeChunk> Chunks);

public sealed record KnowledgeSearchHit(
    string DocumentId,
    string ChunkId,
    string SourceUri,
    string Title,
    int Ordinal,
    string Excerpt,
    double Score)
{
    public string? Locator =>
        KnowledgeChunkLocator.TryGetLocator(ChunkId, out var locator) ? locator : null;
}

public sealed record KnowledgeStoreHealth(
    bool DatabaseReady,
    bool Fts5Ready,
    string SQLiteVersion,
    string DatabasePath);

public sealed record EmbeddingIndexStatus(
    int TotalChunks,
    int IndexedChunks)
{
    public int PendingChunks => Math.Max(0, TotalChunks - IndexedChunks);
}

public sealed record EmbeddingIndexProgress(
    int CompletedChunks,
    int TotalChunks,
    string ChunkId);
